charms/layers/jupyter-notebook/reactive/jupyter_notebook.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+# Copyright (C) 2016  Ghent University
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# pylint: disable=c0111,c0103,c0301
+import subprocess
+
+from charmhelpers.core import templating, hookenv, host
+from charmhelpers.core.hookenv import charm_dir, open_port, status_set
+from charms.reactive import hook, when, when_not, set_state
+
+import charms.apt #pylint: disable=e0611,e0401
+
+
+@hook('upgrade-charm')
+def upgrade_charm():
+    hookenv.log("Upgrading Notebook Charm")
+    try:
+        subprocess.check_call(['pip','install','-U','jupyter'])
+    except subprocess.CalledProcessError as exception:
+        hookenv.log(exception.output)
+    
+@when_not('apt.installed.python-pip', 'jupyter-notebook.installed')
+def pre_install():
+    hookenv.log("Pre-install")
+    charms.apt.queue_install(['python-pip','python3-pip'])#pylint: disable=e1101
```

Comment:
It's best to specify the packages in `layer.yaml`. The java layer is an exception because the installed package depends on the configuration. Example: https://github.com/IBCNServices/tengu-charms/blob/master/charms/layers/sojobo-api/layer.yaml